python/semantic_kernel/connectors/ai/anthropic/services/anthropic_chat_completion.py

```diff
@@ -211,40 +187,88 @@ def _prepare_chat_history_for_request(
             chat_history: The chat history to prepare.
             role_key: The key name for the role/author.
             content_key: The key name for the content/message.
+            stream: Whether the request is for a streaming chat.
 
         Returns:
             A tuple containing the prepared chat history and the first SYSTEM message content.
         """
         system_message_content = None
-        remaining_messages = []
-
+        remaining_messages: list[dict[str, Any]] = []
         system_message_found = False
         for message in chat_history.messages:
             # Skip system messages after the first one is found
             if message.role == AuthorRole.SYSTEM:
                 if not system_message_found:
                     system_message_content = message.content
                     system_message_found = True
-                continue
+            elif message.role == AuthorRole.TOOL:
```

Comment:
I'm a bit confused about what is going on here. Why is there so much manual work needed here to build up the messages list from the chat history? 

With tool calls, the chat history would contain something like:
- User Request (e.g., "What is 1+1?") - AuthorRole.USER
- Assistant Response with FunctionCallContent (contains a tool call to execute) - AuthorRole.ASSISTANT
- FunctionResultContent - AuthorRole.TOOL
- Assistant Response with final answer as TextContent - AuthorRole.ASSISTANT

What does `if tool result message isn't the most recent message, add it to the remaining messages` mean and how would we get into this situation?

Maybe if you could please point me to some Anthropic docs explaining how the messages need to be formed would help my confusion. Thanks!

Reply:
User message is straight forward, agreed.

The main "challenges" here: 
- the FunctionResultContent and how Anthropic only supports "user" and "assistant" roles.
- FunctionCallContent maps to Anthropic ToolUseBlock and TextContent maps to the Anthropic TextBlock. In the initial assistant response, there is often both with the TextContent block representing the chain of thought message https://docs.anthropic.com/en/docs/build-with-claude/tool-use#chain-of-thought in addition to the FunctionCall content
- for tool call results, anthropic expects a "user" message with results populated as such:
```python
tool_results_message["content"].append({
                            "type": "tool_result",
                            "tool_use_id": item.id,
                            content_key: str(item.result),
                        })
```

If there is a tool result message with the SK abstraction, a `tool_results_message` must then be added to the chat history. the results can then be populated into the "content"

Reply:
Appreciate the added context.